src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -2028,4 +2028,100 @@ BOOST_AUTO_TEST_CASE(GradientReparametrizationLayerBetaTest)
   BOOST_REQUIRE_LE(CheckGradient(function), 1e-4);
 }
 
+/**
+ * Simple sequential module residual block test.
+ */
+BOOST_AUTO_TEST_CASE(SimpleSequentialLayerResidualBlockTest)
+{
+  arma::mat outputA /* sequential */, outputB /* residual */, input,
+      deltaA /* sequential */, deltaB /* residual */;
+  Sequential<>* sequential = new Sequential<>(true, false);
+  Sequential<>* residual = new Sequential<>(true, true);
+
+  Linear<>* linearA = new Linear<>(10, 10);
+  linearA->Parameters().randu();
+  linearA->Reset();
+  Linear<>* linearB = new Linear<>(10, 10);
+  linearB->Parameters().randu();
+  linearB->Reset();
+
+  // Add the same layers(with the same parameters) to both normal Sequential and
+  // residual Sequential object.
+  sequential->Add(linearA);
+  sequential->Add(linearB);
+
+  residual->Add(linearA);
+  residual->Add(linearB);
+
+  // Test the Forward function(Pass the same input to both).
+  input = arma::randu(10, 1);
+  sequential->Forward(std::move(input), std::move(outputA));
+  residual->Forward(std::move(input), std::move(outputB));
+
+  CheckMatrices(outputA, outputB - input);
+
+  // Test the Backward function(Pass the same error to both).
+  sequential->Backward(std::move(input), std::move(input), std::move(deltaA));
+  residual->Backward(std::move(input), std::move(input), std::move(deltaB));
+
+  CheckMatrices(deltaA, deltaB - input);
+
+  delete sequential;
+  delete residual;
+  delete linearA;
+  delete linearB;
```

Comment:
I think it should be possible to implement this test without using pointers, can you try exploring that, so that it remains consistent with the rest of the code?

Reply:
hm, I can't comment on the comment below:

> @akhandait Is the required change substantial? @zoq I'd like to know your take on this. I think it's better if we use lesser workarounds.

I think at this stage the solution is fine and necessary, if we can find a better solution later e.g. by using ``unique_ptr`` we can remove the part.

Reply:
@akhandait Okay, do you mind opening an issue for this, so that we keep track of this change?

Reply:
Sure, I think that makes sense in this case. I also think we should get this merged for now and have the issue as a platform for further change.